common/comms/src/inbound_message_service/message_dispatcher.rs

```diff
@@ -0,0 +1,146 @@
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use derive_error::Error;
+use std::collections::HashMap;
+
+#[derive(Debug, Error)]
+pub enum DispatchError {
+    /// A dispatch route was not defined for the specific message type
+    MessageHandlerUndefined,
+}
+
+/// Each message struct that will be dispatched should implement Dispatchable. The dispatch_type function is used by the
+/// MessageDispatcher to be able to distinguish between different message types and select which handler function should
+/// process the message.
+pub trait Dispatchable {
+    fn dispatch_type(&self) -> u32;
+}
+
+/// Format required of handler functions specified by dispatch routes
+type HandlerFunctionFormat<DispMsg> = fn(msg_data: DispMsg) -> Result<(), DispatchError>;
+
+#[derive(Clone, Debug)]
+pub struct MessageDispatcher<DispMsg: Dispatchable> {
+    handlers: HashMap<u32, HandlerFunctionFormat<DispMsg>>,
+}
+
+impl<DispMsg> MessageDispatcher<DispMsg>
+where DispMsg: Dispatchable
+{
+    /// Construct a new MessageDispatcher with no defined dispatch routes
+    pub fn new() -> MessageDispatcher<DispMsg> {
+        MessageDispatcher {
+            handlers: HashMap::new(),
+        }
+    }
+
+    /// This function allows a new dispatch route to be specified and added to the handlers, all received messaged that
+    /// are of the dispatch type will be routed to the specified handler_function
+    pub fn route(mut self, dispatch_type: u32, handler_function: HandlerFunctionFormat<DispMsg>) -> Self {
+        self.handlers.insert(dispatch_type, handler_function);
+        self
+    }
+
+    /// This function can be used to forward a message to the correct function handler
+    pub fn dispatch(&self, msg_data: DispMsg) -> Result<(), DispatchError> {
+        match self.handlers.get(&msg_data.dispatch_type()) {
+            Some(dispatch_function) => {
+                dispatch_function(msg_data)?;
+                Ok(())
+            },
+            None => Err(DispatchError::MessageHandlerUndefined),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_route_and_dispatch() {
+        #[derive(PartialEq, Debug)]
+        pub enum DispatchType {
```

Comment:
Since Dispatch types have to be globally unique and known by the system, we may as well push this idea into the main library and build a nice readable registry of message types.

Reply:
Yes, when we do the do the domain message types then a nice readable registry is a good idea.